### PR TITLE
Replace SortedFractionMedianList with SortedLinkedListWithMedian

### LINKED ISSUE #134: Devs SBAT reuse code with SortedFractionMedianList
SortedFractionMedianList  should inherit from SortedLinkedList or AddressSortedLinkedList

### PR DESCRIPTION
**Pull request by asaj**
### Description

This PR removes the last of the independent linked list implementations, adding `SortedLinkedListWithMedian` which inherits from `SortedLinkedList`.

To avoid using fractions, which require an extra storage slot and necessitate their own comparators, `SortedOracles` stores all oracle reports as fixed point fractions with a denominator of 2 ^ 64.

### Tested

Unit tests

### Other changes
- Fix minor race condition causing `BondedDeposits` test flakiness

### Related issues

- Fixes #134 

### Backwards compatibility

Not backwards compatible as storage schema has changed

### CODE REVIEW

File: packages/protocol/contracts/common/linkedlists/AddressSortedLinkedListWithMedian.sol
Comment (m-chrzan) [inline packages/protocol/contracts/common/linkedlists/AddressSortedLinkedListWithMedian.sol:16]: Why is the bit shift necessary? Can't we store the address in the lower bytes?

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (m-chrzan) [inline packages/protocol/contracts/stability/SortedOracles.sol:44]: I guess this is a dual comment to yours on my PR: do we need to log denominators if they're constant?

File: packages/protocol/contracts/common/linkedlists/SortedLinkedListWithMedian.sol
Comment (yorhodes) [inline packages/protocol/contracts/common/linkedlists/SortedLinkedListWithMedian.sol:94]: Why does it make sense to zero the median on a `remove()` when the list is empty?
Should control flow continue to `updateMedian(emptyList, none)` and `emptyList.remove(key)`?

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (yorhodes) [inline packages/protocol/contracts/stability/SortedOracles.sol:197]: Will this lead to inadvertent divide by zero errors?

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (yorhodes) [inline packages/protocol/contracts/stability/SortedOracles.sol]: It's a little unpleasant that consumers (like `SortedOracles`) of the libraries have to worry about this abstraction depth
Not immediately clear there is a workaround

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (yorhodes) [inline packages/protocol/contracts/stability/SortedOracles.sol:262]: +1 for explicit semantics

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (yorhodes) [inline packages/protocol/contracts/stability/SortedOracles.sol:278]: Ditto here, it seems to me like behavior around empty should be embedded rather than signalled in logs in this way

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (yorhodes) [inline packages/protocol/contracts/stability/SortedOracles.sol:128]: +1, this always made me nervous

File: packages/protocol/migrationsConfig.js
Comment (yorhodes) [inline packages/protocol/migrationsConfig.js:87]: Should test contracts live in a separate migrations config?
On the flipside might be dangerous to separate test/prod migrations

**Review by yorhodes — DISMISSED** [review]
A few comments around handling empty lists

File: packages/protocol/contracts/common/linkedlists/SortedLinkedListWithMedian.sol
Comment (asaj) [reply packages/protocol/contracts/common/linkedlists/SortedLinkedListWithMedian.sol:94]: I think we just want some way of detecting that there is no median, and setting it to the zero key, which is invalid, is the only way I can think of to do that.

This line was supposed to check if `numElements == 1`, oops!

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (asaj) [reply packages/protocol/contracts/stability/SortedOracles.sol:197]: It will.  It's supposed to be invalid for there to be zero oracle rates reported. Any suggestions for how to enforce this?

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (asaj) [reply packages/protocol/contracts/stability/SortedOracles.sol]: It should have used `getNumElements()`, done

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (asaj) [reply packages/protocol/contracts/stability/SortedOracles.sol:278]: Agreed. It's invalid for SortedOracles to not return an exchange rate. Any suggestions for what to do here?

File: packages/protocol/migrationsConfig.js
Comment (asaj) [reply packages/protocol/migrationsConfig.js:87]: I think this is fine since we don't actually migrate the test contract in our migrations.

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (yorhodes) [reply packages/protocol/contracts/stability/SortedOracles.sol:278]: Currently `removeExpiredReports` is the only caller of `removeReport` and requires that the number to remove is strictly less than the number of reports.

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (yorhodes) [reply packages/protocol/contracts/stability/SortedOracles.sol:197]: Upon initialization insert 1 report and prevent removals which decrease number of elements below 1?

File: packages/protocol/contracts/common/linkedlists/SortedLinkedListWithMedian.sol
Comment (yorhodes) [reply packages/protocol/contracts/common/linkedlists/SortedLinkedListWithMedian.sol:94]: I think `numElements == 1` => `median = 0` is general enough. We should require at the top of this function that `numElements > 0` though.

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (yorhodes) [reply packages/protocol/contracts/stability/SortedOracles.sol]: Don't see this change checked in

**Review by yorhodes — COMMENTED** [review]
Followup comments

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (asaj) [reply packages/protocol/contracts/stability/SortedOracles.sol]: Forgot to push, done

**Comment by asaj** [comment]
@yorhodes and I talked offline, decided to decouple the issue of no reports from this PR

**Review by yorhodes — DISMISSED** [review]
Approving with the caveat that `SortedOracles` should maintain the invariant of at least 1 report more explicitly. `SortedLinkedListWithMedian` should generally allow empty lists.

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (m-chrzan) [inline packages/protocol/contracts/stability/SortedOracles.sol:18]: What do you think about waiting for Fixidity to be merged and use it here, to not add another ad-hoc fixed fraction implementation?

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (asaj) [reply packages/protocol/contracts/stability/SortedOracles.sol:44]: IMO it's a nice touch since it makes the log message more readable

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (asaj) [reply packages/protocol/contracts/stability/SortedOracles.sol:18]: Will fixidity work out of the box with this? Would rather get this in sooner for audit purposes, happy to move to fixidity in a follow-up PR

Comment (m-chrzan) [reply packages/protocol/contracts/stability/SortedOracles.sol:18]: It should. Comparisons on Fixidity fractions are the same as on integers.

What else could be a concern?

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (m-chrzan) [reply packages/protocol/contracts/stability/SortedOracles.sol:44]: The extra readability feels nice, but it is redundant information, and this is at a super low level of the stack. What do you think about suffixing the event fields with `Fixed` or `Fraction` to make it clear that we're not talking about the integer value?

File: packages/protocol/contracts/stability/SortedOracles.sol
Comment (asaj) [reply packages/protocol/contracts/stability/SortedOracles.sol:18]: Values are uint256's for all of these implementations

File: packages/protocol/contracts/common/linkedlists/AddressSortedLinkedListWithMedian.sol
Comment (asaj) [reply packages/protocol/contracts/common/linkedlists/AddressSortedLinkedListWithMedian.sol:16]: This is consistent with the way we did it in other implementations

**Review by yorhodes — APPROVED** [review]
reapprove

**Pull request merged**
